### PR TITLE
seccomp: add support for flags

### PR DESCRIPTION
**Pull request by alban**
List of seccomp flags defined in runtime-spec:
* SECCOMP_FILTER_FLAG_TSYNC
* SECCOMP_FILTER_FLAG_LOG
* SECCOMP_FILTER_FLAG_SPEC_ALLOW

Note that runc does not apply SECCOMP_FILTER_FLAG_TSYNC. It does not
make sense to apply the seccomp filter on only one thread; other threads
will be terminated after exec anyway.

See similar commit in crun:
https://github.com/containers/crun/commit/fefabffa2816ea343068ed036a86944393db189a

Note that SECCOMP_FILTER_FLAG_WAIT_KILLABLE_RECV (introduced by
https://git.kernel.org/pub/scm/linux/kernel/git/kees/linux.git/commit/?id=c2aa2dfef243
in Linux 5.19-rc1) is not added yet because Linux 5.19 is not released
yet.

-----

Tests:
```
$ make localintegration
...
ok 103 runc run [seccomp] (SECCOMP_FILTER_FLAG_*)
```

cc @rata @saschagrunert

### CODE REVIEW

File: tests/integration/seccomp.bats
Comment (rata) [inline tests/integration/seccomp.bats:80]: I'd add a comment saying we just test that the flag is accepted and the container runs as expected. But we don't read the logs

**Review by rata — DISMISSED** [review]
I'm no maintainer here, but this LGTM. Just a minor comment on the integration test

File: tests/integration/seccomp.bats
Comment (alban) [reply tests/integration/seccomp.bats:82]: Thanks. Updated.

**Review by kolyshkin — CHANGES_REQUESTED** [review]
> I didn't add support for `SECCOMP_FILTER_FLAG_SPEC_ALLOW` (#2430) because libseccomp-golang does not support it yet.

It does; calling `filter.SetSSB(true)` will enable `SECCOMP_FILTER_FLAG_SPEC_ALLOW`.

File: libcontainer/specconv/spec_linux.go
Comment (kolyshkin) [inline libcontainer/specconv/spec_linux.go:1025]: It would be nice to not have this set of flags handled in two separate places. Perhaps we can add some common part to `libcontainer/seccomp`?

While at it, adding `SCMP_FLTATR_CTL_OPTIMIZE` would be awesome

Comment (saschagrunert) [reply libcontainer/specconv/spec_linux.go:1025]: Right now we have four supported flags following the same naming convention supported by the SECCOMP_SET_MODE_FILTER operation.

SECCOMP_FILTER_FLAG_LOG
SECCOMP_FILTER_FLAG_NEW_LISTENER
SECCOMP_FILTER_FLAG_SPEC_ALLOW
SECCOMP_FILTER_FLAG_TSYNC

How would `SCMP_FLTATR_CTL_OPTIMIZE` map to that?

Comment (alban) [reply libcontainer/specconv/spec_linux.go:1025]: > While at it, adding `SCMP_FLTATR_CTL_OPTIMIZE` would be awesome

The [set of flags listed in the runtime-spec](https://github.com/opencontainers/runtime-spec/blob/8958f93039ab90be53d803cd7e231a775f644451/config-linux.md) is a [subset](https://github.com/opencontainers/runtime-spec/pull/1096) of the flag supported by the Linux API in the seccomp syscall:

> flags (array of strings, OPTIONAL) - list of flags to use with seccomp(2).
>
> A valid list of constants is shown below.
>
> SECCOMP_FILTER_FLAG_TSYNC
> SECCOMP_FILTER_FLAG_LOG
> SECCOMP_FILTER_FLAG_SPEC_ALLOW

`SCMP_FLTATR_CTL_OPTIMIZE` is part of libseccomp API but not part of the Linux API. It is not in the runtime-spec because container runtimes don't have to use libseccomp. Are you suggesting runc should support it even if it's not in the spec?

Comment (alban) [reply libcontainer/specconv/spec_linux.go]: > It would be nice to not have this set of flags handled in two separate places. Perhaps we can add some common part to libcontainer/seccomp?

Now that I have added the 2 additional flags, the code looks different, and I don't see how to add some common part to libcontainer/seccomp... any suggestions?

File: libcontainer/seccomp/seccomp_linux.go
Comment (saschagrunert) [inline libcontainer/seccomp/seccomp_linux.go:92]: Note: I think we should have them as consts in the runtime spec. I proposed that change in https://github.com/opencontainers/runtime-spec/pull/1138

**Review by saschagrunert — DISMISSED** [review]

File: tests/integration/seccomp.bats
Comment (kolyshkin) [inline tests/integration/seccomp.bats:120]: I don't like the same code being repeated 3 times. Can probably factor it out to a function accepting the list of flags. That way we can also test a combination of flags, not just individual ones.

File: libcontainer/specconv/spec_linux.go
Comment (kolyshkin) [reply libcontainer/specconv/spec_linux.go:1025]: `SECCOMP_FILTER_FLAG_TSYNC` is there already and it's on by default (in libseccomp-golang). How would it be supported? Do you mean "let's explicitly ignore it", as I see no other way.

File: libcontainer/specconv/spec_linux.go
Comment (kolyshkin) [reply libcontainer/specconv/spec_linux.go]: Re `SCMP_FLTATR_CTL_OPTIMIZE`.

Perhaps runc should just set it to `2` unconditionally. Also, this is orthogonal to this PR, so nevermind.

File: libcontainer/specconv/spec_linux.go
Comment (kolyshkin) [inline libcontainer/specconv/spec_linux.go:1026]: `SECCOMP_FILTER_FLAG_TSYNC` can just be silently ignored, i.e. it does not make sense to pass it around.

File: libcontainer/configs/config.go
Comment (kolyshkin) [inline libcontainer/configs/config.go:36]: So, this is our internal representation. Not sure why we use strings here at all. Could be
```go
const SecccompFilterFlags (
	SeccompTSync = iota
	SeccompFlagLog
	SeccompFlagSpecAllow
)
```
and then specconv can do the conversion.

Yet better, do that in libseccomp-golang can have these. I'll take a look.

File: libcontainer/seccomp/seccomp_linux.go
Comment (kolyshkin) [reply libcontainer/seccomp/seccomp_linux.go:92]: Note that seccomp-golang sets this flag unconditionally, and we (runc/libcontainer/seccomp/patchbpf) do not set this at all (not sure if this was by design or just forgotten; @cyphar WDYT?)

From what I see, it seems that for runc it does not make sense to do so, as we execute runc init right after applying seccomp rules, and it does not matter much whether these rules were applied to all threads or not.

File: libcontainer/configs/config.go
Comment (alban) [reply libcontainer/configs/config.go:36]: Did you have a chance to look at libseccomp-golang?

File: tests/integration/seccomp.bats
Comment (alban) [reply tests/integration/seccomp.bats]: Updated with a loop to test all set of flags possible

File: libcontainer/seccomp/seccomp_linux.go
Comment (cyphar) [reply libcontainer/seccomp/seccomp_linux.go:92]: There was an issue with setting it in patchbpf, I can't remember the exact details.

File: tests/integration/seccomp.bats
Comment (kolyshkin) [inline tests/integration/seccomp.bats:93]: This sentence is not finished.

Comment (alban) [reply tests/integration/seccomp.bats]: Fixed

**Comment by vnsavage** [comment]
It seems like all requested changes are implemented on this diff? Is there anything else that needs doing before this gets merged?

File: tests/integration/seccomp.bats
Comment (AkihiroSuda) [inline tests/integration/seccomp.bats:88]: Needs "SCMP_ARCH_AARCH64","SCMP_ARCH_ARM"

**Review by AkihiroSuda — APPROVED** [review]

**Comment by alban** [comment]
It seems the last changes reliably break `TestUsernsCheckpoint`. Any ideas why?

**Comment by kolyshkin** [comment]
The error is EINVAL from move_mount(2) syscall. It can be returned if the kernel is too old, but it seems that
 - Linux v5.15 is used, which is sufficient;
 - criu does the check for the `MOVE_MOUNT_SET_GROUP` feature and finds it working.

There are other cases when the move mount set group can return EINVAL, but it does not seem likely.

Ah! It might be caused by the fact that criu ppa comes with criu 3.17-1ppa2.20.04 which is old (should be 3.17.1). I just saw that @rst0git updated that package 3 hours ago. So, let me restart the CI.

**Comment by kolyshkin** [comment]
TestUsernsCheckpoint also fails on CentOS Stream 9, although with a different diagnostics (from https://cirrus-ci.com/task/5426803694108672?logs=unit_tests#L1368):

```
    checkpoint_test.go:163: (00.041591) pie: 1: restoring lsm profile (current) unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
    checkpoint_test.go:163: (00.041654) Error (criu/cr-restore.c:1510): 65454 stopped by signal 11: Segmentation fault
    checkpoint_test.go:163: (00.042114) mnt: Switching to new ns to clean ghosts
    checkpoint_test.go:163: (00.042270) uns: calling exit_usernsd (-1, 1)
    checkpoint_test.go:163: (00.042312) uns: daemon calls 0x47fe40 (65452, -1, 1)
    checkpoint_test.go:163: (00.042319) uns: `- daemon exits w/ 0
    checkpoint_test.go:163: (00.042824) uns: daemon stopped
    checkpoint_test.go:163: (00.042829) Error (criu/cr-restore.c:2536): Restoring FAILED.
    checkpoint_test.go:163: === END ===
```

**Comment by kolyshkin** [comment]
@alban you also need to fix a linter warning

File: libcontainer/specconv/spec_linux.go
Comment (kolyshkin) [inline libcontainer/specconv/spec_linux.go:1031]: nit: `s/flags/flag/`

**Comment by alban** [comment]
* PR updated and rebased.
* The linter warning is fixed.
* The TestUsernsCheckpoint issue is resolved in github-actions but there is still an issue in cirrus-ci with TestUsernsCheckpoint and TestCheckpoint.

**Comment by rst0git** [comment]
> TestUsernsCheckpoint also fails on CentOS Stream 9, although with a different diagnostics (from https://cirrus-ci.com/task/5426803694108672?logs=unit_tests#L1368):

We recently updated the criu package in CentOS Stream (https://github.com/opencontainers/runc/issues/3532)

**Review by kolyshkin — APPROVED** [review]
LGTM (CI failure is unrelated and is being addressed in https://github.com/opencontainers/runc/issues/3532)

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
> Required statuses must pass before merging

Can't hit the merge button :(

**Comment by rata** [comment]
Again `--- FAIL: TestUsernsCheckpoint (0.43s)`. I guess @alban needs to push again (or maybe rebase) so the tests are happy? Or is the problem on the CI not fixed?

**Comment by rata** [comment]
@alban a fix was just merged for the test (https://github.com/opencontainers/runc/pull/3539). If you rebase and push, the tests should re-start and work this time :)

**Comment by alban** [comment]
@rata updated!

**Review by kolyshkin — APPROVED** [review]
LGTM

**Review by AkihiroSuda — APPROVED** [review]

**Pull request merged**
